### PR TITLE
Arithmetic

### PR DESCRIPTION
**Pull request by sujaymansingh**
https://sites.google.com/site/prologsite/prolog-problems/2

### CODE REVIEW

File: lib/arithmetic.ex [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+  @doc """
+  Given a number n, return the smallest prime p such that p > n
+
+    iex> Arithmetic.next_prime(2)
+    3
+    iex> Arithmetic.next_prime(5000)
+    5003
+  """
+  defmemo next_prime(n) do
```
Comment (samsondav) [inline lib/arithmetic.ex:35]: OMG very nice, didn't know about this!

File: lib/arithmetic.ex [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  def prime_factors(n) do
+    prime_factors(n, 2, [])
+  end
+
+  defp prime_factors(1, _, acc), do: Lists.reverse(acc)
+
+  defp prime_factors(n, p, acc) do
+    remainder = rem(n, p)
+
+    if remainder == 0 do
```
Comment (samsondav) [inline lib/arithmetic.ex]: I feel like this could be factored out into a function with two clauses?

File: lib/arithmetic.ex [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+  Construct a list containing the prime factors and their multiplicity.
+
+    iex> Arithmetic.prime_factors_mult(315)
+    [{3, 2}, {5, 1}, {7, 1}]
+  """
+  def prime_factors_mult(n) do
+    n
+    |> prime_factors
+    |> Lists.encode()
```
Comment (samsondav) [inline lib/arithmetic.ex:80]: What does Lists.encode do?

**Review by samsondav — APPROVED** [review]
Super cool @sujaymansingh and your elixir is looking impeccable.

File: lib/arithmetic.ex (hunk 2 above)
Comment (sujaymansingh) [reply lib/arithmetic.ex]: Ah good point @samphilipd. I didn't realise that you could have `rem(x, y) == 0` in a guard!

File: lib/arithmetic.ex (hunk 3 above)
Comment (sujaymansingh) [reply lib/arithmetic.ex:80]: It's a reference to a function in the previous exercise.

```ex
iex> Lists.encode([:a, :a, :a, :a, :b, :c, :c, :a, :a, :d, :e, :e, :e, :e])
[{4, :a}, {1, :b}, {2, :c}, {2, :a}, {1, :d}, {4, :e}]
```

It's not entirely clear what `encode` actually does, but I've been trying to keep the names of the functions the same as what the exercises ask for (even when they are vague :))

**Comment by sujaymansingh** [comment]
Thanks again @samphilipd !

**Pull request merged**